include/dice/template-library/ranges.hpp

```diff
@@ -2,40 +2,466 @@
+				}
+			}
+
+			template<std::ranges::input_range R>
+			friend constexpr auto operator|(R &&range, const empty_fn &self)
+					-> decltype(self(std::forward<R>(range))) {
+				return self(std::forward<R>(range));
+			}
+		};
+
+		struct non_empty_fn {
+			template<std::ranges::input_range R>
+			constexpr bool operator()(R &&range) const {
+				// Implemented as the negation of empty_fn for code reuse.
+				return !empty_fn{}(std::forward<R>(range));
+			}
+
+			template<std::ranges::input_range R>
+			friend constexpr auto operator|(R &&range, const non_empty_fn &self)
+					-> decltype(self(std::forward<R>(range))) {
+				return self(std::forward<R>(range));
+			}
+		};
+	}// namespace ranges_algo_detail
+
+	/**
+     * @brief Checks if a range is empty.
+     *
+     * This is a pipeable, terminal algorithm. It uses `std::ranges::empty` if available,
+     * otherwise it falls back to comparing `cbegin()` and `cend()`.
+     *
+     * @return A closure that returns true if the range is empty otherwise false.
+     */
+	inline constexpr ranges_algo_detail::empty_fn empty{};
+
+	/**
+     * @brief Checks if a range is not empty.
+     *
+     * This is a pipeable, terminal algorithm and the logical negation of `dtl::empty`.
+     *
+     * @return A closure that returns true if the range has at least one element otherwise false.
+     */
+	inline constexpr ranges_algo_detail::non_empty_fn non_empty{};
+
+}// namespace dice::template_library
+
+// remove_element adaptor
+namespace dice::template_library {
+	namespace ranges_algo_detail {
+		template<typename T, typename Pred>
+		struct remove_element_fn {
+			T value_;
+			Pred pred_;
+
+			template<typename V, typename P>
+			constexpr remove_element_fn(V &&value, P &&pred)
+				: value_(std::forward<V>(value)), pred_(std::forward<P>(pred)) {}
+
+			template<std::ranges::viewable_range R>// viewable_range ensures there are no views into dangling stuff
+			constexpr auto operator()(R &&range) const {
+				return std::forward<R>(range) | std::views::filter([value = value_, pred = pred_](const auto &element) {
+						   return !std::invoke(pred, element, value);
+					   });
+			}
+
+			template<std::ranges::viewable_range R>// viewable_range ensures there are no views into dangling stuff
+			friend constexpr auto operator|(R &&range, const remove_element_fn &self)
+					-> decltype(self(std::forward<R>(range))) {
+				return self(std::forward<R>(range));
+			}
+		};
+	}// namespace ranges_algo_detail
+
+	/**
+	 * Creates a lazy, pipeable view of a range that excludes elements matching a given value. This adaptor does not modify the source range.
+	 *
+	 * @param value The value to filter out of the source range.
+	 * @param pred Optional binary predicate for comparison. Defaults to `std::ranges::equal_to`.
+	 * @return A range adaptor closure.
+	 */
+	template<typename T, typename Pred = std::ranges::equal_to>
+	constexpr auto remove_element(T &&value, Pred &&pred = Pred{}) {
```

Comment:
I think `T const &` should be sufficient

Reply:
discussed: we'll keep this